vpr/src/device/rr_graph_view.h

```diff
@@ -0,0 +1,99 @@
+ * - node_ptc_storage
+ * - node_fan_in_storage
+ * - rr_node_indices
+ *
+ * Note that the RRGraphView does not own the storage
+ * It serves a virtual read-only protocol for
+ * - placer
+ * - router
+ * - timing analyzer
+ * - GUI
+ *
+ * Note that each client of rr_graph may get a frame view of the object
+ * The RRGraphView is the complete frame view of the routing resource graph
+ * - This helps to reduce the memory footprint for each client
+ * - This avoids massive changes for each client on using the APIs
+ *   as each frame view provides adhoc APIs for each client
+ *
+ * TODO: more compact frame views will be created, e.g., 
+ * - a mini frame view: contains only node and edges, representing the connectivity of the graph
+ * - a geometry frame view: an extended mini frame view with node-level attributes, 
+ *                          in particular geometry information (type, x, y etc).
+ *
+ */
+class RRGraphView {
+    /****************
+     * Constructors
+     ****************/
+  public:
+    RRGraphView();
+
+    /****************
+     * Accessors
+     ****************/
+  public:
+    /* Get the type of a routing resource node */
+    t_rr_type node_type(const RRNodeId& node) const;
+
+    /* Returns the index of the specified routing resource node.  (x,y) are
+     * the location within the FPGA, rr_type specifies the type of resource,
+     * and ptc gives the number of this resource.  ptc is the class number,
+     * pin number or track number, depending on what type of resource this
+     * is.  All ptcs start at 0 and go up to pins_per_clb-1 or the equivalent.
+     * There are class_inf size SOURCEs + SINKs, type->num_pins IPINs + OPINs,
+     * and max_chan_width CHANX and CHANY (each).
+     *
+     * Note that for segments (CHANX and CHANY) of length > 1, the segment is
+     * given an rr_index based on the (x,y) location at which it starts (i.e.
+     * lowest (x,y) location at which this segment exists).
+     * This routine also performs error checking to make sure the node in
+     * question exists.
+     *
+     * The 'side' argument only applies to IPIN/OPIN types, and specifies which
+     * side of the grid tile the node should be located on. The value is ignored
+     * for non-IPIN/OPIN types
+     */
+    RRNodeId find_node(const int& x,
+                       const int& y,
+                       const t_rr_type& type,
+                       const int& ptc,
+                       const e_side& side = NUM_SIDES) const;
+
+    /****************
+     * Mutators
+     ****************/
+  public:
+    /* The ONLY mutators allowed */
+    void set_internal_data(t_rr_graph_storage* node_storage,
+                           t_rr_node_indices* rr_node_indices);
+
+    /****************
+     * internal data storage, can be accessed by parent class RRGraphBuilderView
+     ****************/
+  protected:
+    /* node-level storage including edge storages */
+    t_rr_graph_storage* node_storage_;
```

Comment:
Strive for making members that are only set _once_ in the constructor, `const`.

```
t_rr_graph_storage* const node_storage_;
t_rr_node_indices* const rr_node_indices_;
```

That way, it is immediately clear for someone reading the code that these values will never be assigned to again, and it helps make it easier to reason about the code.

this also means, that you will have to have a
`RRGraphView(t_rr_graph_storage* , t_rr_node_indices*)` constructor that initializes via the initializer list, because otherwise assignment will not be possible anymore aftewards: So e.g. this works

```
RRGraphView(t_rr_graph_storage* storage, t_rr_node_indices* indices) 
    : node_storage_(storage), rr_node_indices_(indices) {}
```

While the following won't, as it attempts to assign things _after_ the initializer list:
```
RRGraphView(t_rr_graph_storage* storage, t_rr_node_indices* indices) {
  node_storage_ = storage;
  rr_node_indices_ = indices;
}
```
So using this technique of making as much `const` as can be (the stuff that will only be set once in the constructor) also lets us adopt the correct habit of initializing as much as possible in the constructor initializer list. And it is easy to find the one place where it happens -> overall improved code quality.

Reply:
Agree. Updated.